### ISSUE TITLE
Cannot delete receiver deployment from webpage

### ISSUE THREAD

**Issue by j-sayers** [MotusDev/Motus-TO-DO#1056]
Project 467 has a duplicate receiver deployment they want deleted, only the admin delete tool isn't working. 

https://motus.org/data/receivers/edit?projectID=467&id=3979

Try to delete deployment 8316, I get the confirmation dialog, but when the page refreshes it's still there.

**Comment by Hussell** [comment]
Tracing this through the code, it seems prDeleteSensorDeployment is silently failing. Possibly there's a new dependency which isn't being handled, e.g. sensor_nodes or det_gps_timeline?

Looking at the two deployments (8313 and 8316), 8316 has more metadata filled in, so I figure it's 8313 which should be deleted. The example link above doesn't work any more, go here instead:
https://motus.org/data/station/deployment/edit?projectID=467&id=8313

Hit the "Delete deployment" button in the upper-right (administrators only). The page reports success because no error is thrown by the procedure, but so far as I can tell nothing gets changed in the database.

**Comment by cjardine-bsc** [comment]
Huh, this is actually an odd case. The problem is a duplicate key in taghits runs. Was there possibly a third deployment at some point? That would explain it. 

I re-ran prTaghitsSummariesSensorID for the sensor and that seems to have solved it. I suspect the job failed in the procs queue which led to this. 

we need a better way to deal with failed procs queue steps....

**Comment by j-sayers** [comment]
Perhaps the same issue occurring here? https://motus.org/data/project/stations?setStationId=9806

I keep "deleting" the receiver deployment but it's still there.

![image](https://user-images.githubusercontent.com/55001144/153436758-e801d066-56d6-454c-a9f6-b4228ad07dbe.png)

**Comment by j-sayers** [comment]
Same as #595?

**Comment by cjardine-bsc** [comment]
You're trying to delete Deployment ID:
8263?

**Comment by j-sayers** [comment]
> You're trying to delete Deployment ID: 8263?

Yes please. And following that, station_id 9806

**Comment by leberrigan** [comment]
@j-sayers are you still unable to delete this deployment?

**Comment by j-sayers** [comment]
I was just able to delete an overlapping deployment of a receiver at the _same_ station. 

The above failed attempts involved overlapping deployments of the same receiver, but at _different_ stations.

**Comment by j-sayers** [comment]
Another case of being unable to delete a receiver deployment https://motus.org/data/station/deployment/edit?projectID=483&id=8770

**Comment by denislepage** [comment]
I tried to run manually the delete proc. It's quite slow, but there is also a conflict in the det_taghits_run table.

```
EXEC prDeleteSensorDeployment 483, 8770

Msg 50000, Level 16, State 1, Procedure prDeleteSensorDeployment, Line 49
Error Occured: Violation of PRIMARY KEY constraint 'PK_det_taghits_runs'. Cannot insert duplicate key in object 'dbo.det_taghits_runs'. The duplicate key value is (2710, 5016, 44102, 0, 0, 5890293, 323150135).

```
The reason being precisely that there are 2 sensor deployment records for the same run, and we are now trying to set both to the same value in the delete proc:

`    UPDATE det_taghits_runs SET sensor_deploy_id = 0 WHERE sensor_deploy_id = 8770`

For example:

` select * from det_taghits_runs where run_id = 323150135 `

I am not sure this is a perfect solution, but a quick fix could be to avoid replicating records :

```

    UPDATE a SET sensor_deploy_id = 0 from det_taghits_runs a WHERE sensor_deploy_id = 8770
    and not exists (select * from det_taghits_runs b where a.run_id = b.run_id and b.sensor_deploy_id = 0)
  
    delete from det_taghits_runs WHERE sensor_deploy_id = 8770 
    
```

I think that my condition is "not exists" is correct (e.g. each run would be expected to be listed once in that table, except when there are overlapping deployments or tags or sensors, or perhaps ambiguous tags also.

I'm not sure this proc is common enough that we should try to optimize it. Adding a new index is always costly, but I can look into it once we have put a fix in place

**Comment by cjardine-bsc** [comment]
Ha. I just pushed a similar change. 

The problem with yours, is that there will be orphaned records with sensor_deploy_id in det_taghits_runs. 

I'm deleting the duplicates instead. 

delete FROM [motus].[dbo].[det_taghits_runs]  
  where sensor_deploy_id = 8770 and exists (select * from det_taghits_runs b where det_taghits_runs.run_id =b.run_id and sensor_deploy_id = 0) 
  
 Did you just run your thing for 8770? If so mine will delete more than it should.

**Comment by denislepage** [comment]
I didn’t run mine.

The second line (delete) in mine would have handled the leftovers by deleting anything left with the sensor ID being deleted.

In your case, you still will have to update the records remaining with the deleted deploy ID.

Same difference, I think, but I’m fine going either way.

**Comment by j-sayers** [comment]
Running into something similar with station 12612 https://motus.org/data/project/stations?setStationId=12612.

They accidentally created several unnecessary deployments which they want deleted. I was able to delete _some_ of them, but for receiver deployments 8743 and 8742 I am running into something very similar where the dialog says they have been deleted, but in fact they have not been.

**Comment by cjardine-bsc** [comment]
The first thing we need to sort out is how users are creating duplicate deployments in the first place. This wasn't possible until after we released beta 10. Have you guys already tested this and created a git? 

Otherwise, this is the same issue as before, just impacting a different summary table. I fixed that, but its not the solution. We need to prevent users from creating overlapping deployments either at the same station, or for the same sensor.

**Comment by j-sayers** [comment]
It's clear that overlapping deployments are being created somehow and we need to get to the bottom of it,

In some cases though, like this most recent one, they just accidentally made too many deployments and needed to delete them.


Josh Sayers
Motus Wildlife Tracking System
Birds Canada | Oiseaux Canada
PO Box 160, 115 Front Rd. Port Rowan, Ontario. N0E 1M0
birdscanada.org | ***@***.******@***.***>
motus.org | ***@***.******@***.***>
519-586-3531 X 117






------ Original Message ------
From "Catherine Jardine" ***@***.******@***.***>>
To "MotusDev/Motus-TO-DO" ***@***.******@***.***>>
Cc "Josh Sayers" ***@***.******@***.***>>; "State change" ***@***.******@***.***>>
Date 2022-06-27 7:04:47 PM
Subject Re: [MotusDev/Motus-TO-DO] Cannot delete receiver deployment from webpage (Issue #1056)


The first thing we need to sort out is how users are creating duplicate deployments in the first place. This wasn't possible until after we released beta 10. Have you guys already tested this and created a git?

Otherwise, this is the same issue as before, just impacting a different summary table. I fixed that, but its not the solution. We need to prevent users from creating overlapping deployments either at the same station, or for the same sensor.

**Comment by cjardine-bsc** [comment]
Closed in deference to #1234

One two three four, dub deployments are no more.
